src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs

```diff
@@ -163,14 +171,115 @@ public static void Main()
 42
 ";
             CompileAndVerify(source, expectedOutput: expected);
+
+            var comp = CreateCompilation([source, RuntimeAsyncAwaitHelpers], targetFramework: TargetFramework.Net90);
+            comp.Assembly.SetOverrideRuntimeAsync();
+
+            var verifier = CompileAndVerify(comp, verify: Verification.Fails with
+            {
+                ILVerifyMessage = ReturnValueMissing("F", "0x2e"),
+            }, symbolValidator: verify);
+            verifier.VerifyDiagnostics();
+
+            verifier.VerifyIL("Test.F", """
+                {
+                  // Code size       47 (0x2f)
+                  .maxstack  3
+                  IL_0000:  call       "System.Threading.Tasks.TaskFactory System.Threading.Tasks.Task.Factory.get"
+                  IL_0005:  ldsfld     "System.Action Test.<>c.<>9__1_0"
+                  IL_000a:  dup
+                  IL_000b:  brtrue.s   IL_0024
+                  IL_000d:  pop
+                  IL_000e:  ldsfld     "Test.<>c Test.<>c.<>9"
+                  IL_0013:  ldftn      "void Test.<>c.<F>b__1_0()"
+                  IL_0019:  newobj     "System.Action..ctor(object, nint)"
+                  IL_001e:  dup
+                  IL_001f:  stsfld     "System.Action Test.<>c.<>9__1_0"
+                  IL_0024:  callvirt   "System.Threading.Tasks.Task System.Threading.Tasks.TaskFactory.StartNew(System.Action)"
+                  IL_0029:  call       "void System.Runtime.CompilerServices.RuntimeHelpers.Await(System.Threading.Tasks.Task)"
+                  IL_002e:  ret
+                }
+                """);
+
+            void verify(ModuleSymbol module)
+            {
+                var test = module.ContainingAssembly.GetTypeByMetadataName("Test");
+                var f = test.GetMethod("F");
+                Assert.Equal(MethodImplOptionsAsync, f.ImplementationAttributes);
+                AssertEx.Equal(["<>c"], test.GetTypeMembers().SelectAsArray(t => t.Name));
+            }
+        }
```

Comment:
Consider verifying that no state machine is produced

Reply:
That's what the `AssertEx.Equal` verifies: the only nested type is the closure for the lambda.